base/scheduler/partr.jl

```diff
@@ -211,7 +154,26 @@ function multiq_deletemin()
     heap = tpheaps[rn1]
     task = heap.tasks[1]
     if ccall(:jl_set_task_tid, Cint, (Any, Cint), task, tid-1) == 0
-        unlock(heap.lock)
+        # This task is stuck to a thread that's likely sleeping, move the task to it's private queue and wake it up
+        # We move this out of the queue to avoid spinning on it
+        tid2 = Threads.threadid(task)
+        if tid2 != 0
+            ntasks = heap.ntasks
+            @atomic :monotonic heap.ntasks = ntasks - Int32(1)
+            heap.tasks[1] = heap.tasks[ntasks]
+            Base._unsetindex!(heap.tasks, Int(ntasks))
+            prio1 = typemax(UInt16)
+            if ntasks > 1
+                multiq_sift_down(heap, Int32(1))
+                prio1 = heap.tasks[1].priority
+            end
+            @atomic :monotonic heap.priority = prio1
+            push!(Base.workqueue_for(tid2), task)
+            unlock(heap.lock)
+            ccall(:jl_wakeup_thread, Cvoid, (Int16,), (tid2 - 1) % Int16)
+        else
```

Comment:
Can you explain why this code is now needed? Maybe it would also make sense to factor it out?

Reply:
I'll add a more comprehensive comment but the gist is:

- Thread 2 with Task1 calls wait for some reason, maybe on a lock
- Thread 2 tries to find more work, fails and goes to sleep (while holding on to task1)
- Thread 1 notifies Task1, scheduling that task
This is where in the new design things might not work as expected, since we're not waking every thread up on every schedule, it can happen that we decide to try and run task1, but it's still stuck to thread 2. This requires waking up thread 2. In order to avoid other tasks from spinning on the stuck task we push it to the private queue of thread 2, since regardless of what happens it has to at least run that task once.

Reply:
In relation to the factoring out, I guess we could factor out a delete task from partr, but not sure if we gain much